File: web/src/lib/template-structure.ts
/**
 * Parse firm template text into structural outline (headings / CREAC).
 * Mirrors services/api/app/services/template_structure.py for client preview.
 */

export type CreacRole =
  | "conclusion"
  | "rule"
  | "explanation"
  | "analysis"
  | "conclusion_close"
  | "introduction"
  | "caption"
  | "other";

export type TemplateSection = {
  id: string;
  label: string;
  role: CreacRole;
  contentExcerpt: string;
  order: number;
  classification?: string;
  slots?: Array<{
    slot_type?: string;
    label?: string;
    replacement_key?: string;
    required?: boolean;
    matched_text?: string;
  }>;
};

const CREAC_PATTERNS: Array<{ re: RegExp; role: CreacRole }> = [
  { re: /^\s*(i+\.?\s+)?conclusion\b/i, role: "conclusion" },
  { re: /^\s*(v+\.?\s+)?conclusion\b/i, role: "conclusion" },
  { re: /^\s*(ii+\.?\s+)?(legal\s+)?standard\b/i, role: "rule" },
  { re: /^\s*(ii+\.?\s+)?rule\b/i, role: "rule" },
  { re: /^\s*(applicable\s+)?law\b/i, role: "rule" },
  { re: /^\s*(iii+\.?\s+)?explanation\b/i, role: "explanation" },
  { re: /^\s*(discussion\s+of\s+(the\s+)?law|legal\s+framework)\b/i, role: "explanation" },
  { re: /^\s*(iv+\.?\s+)?(argument|analysis|application|discussion)\b/i, role: "analysis" },
  { re: /^\s*(totality|discretionary\s+factors|positive\s+equities)\b/i, role: "analysis" },
  { re: /^\s*(i+\.?\s+)?introduction\b/i, role: "introduction" },
  { re: /^\s*(caption|in\s+the\s+matter\s+of)\b/i, role: "caption" },
  { re: /^\s*(certificate\s+of\s+service|table\s+of\s+contents)\b/i, role: "other" },
];

const NUMBERED_HEADING =
  /^\s*((?:[IVXLC]+\.|[A-Z]\.|§?\d+(?:\.\d+)*\.?|[0-9]+(?:\.[0-9]+)*\.?)\s+)(.{3,120})$/;
const ALL_CAPS_HEADING = /^[A-Z0-9][A-Z0-9\s\-–—,.'()]{2,100}$/;
const MARKDOWN_HEADING = /^#{1,3}\s+(.+)$/;

function slug(label: string, order: number): string {
  const base =
    label
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-|-$/g, "")
      .slice(0, 48) || "section";
  return `${base}-${order}`;
}

export function detectCreacRole(label: string): CreacRole {
  const cleaned = (label || "").trim();
  const stripped = cleaned.replace(/^(?:[IVXLC]+\.|[A-Z]\.|§?\d+(?:\.\d+)*\.?)\s*/i, "");
  for (const { re, role } of CREAC_PATTERNS) {
    if (re.test(stripped) || re.test(cleaned)) return role;
  }
  return "other";
}

function isHeadingLine(line: string): boolean {
  const text = line.trim();
  if (!text || text.length > 140) return false;
  if (MARKDOWN_HEADING.test(text) || text.startsWith("## ") || text.startsWith("# ")) return true;
  if (NUMBERED_HEADING.test(text)) return true;
  if (ALL_CAPS_HEADING.test(text) && text.includes(" ") && !text.endsWith(".")) {
    const letters = [...text].filter((c) => /[a-zA-Z]/.test(c));
    if (letters.length >= 4 && letters.every((c) => c === c.toUpperCase())) return true;
  }
  if (detectCreacRole(text) !== "other" && text.split(/\s+/).length <= 8) return true;
  return false;
}

function headingLabel(line: string): string {
  const text = line.trim();
  const md = text.match(MARKDOWN_HEADING);
  if (md?.[1]) return md[1].trim();
  if (text.startsWith("## ")) return text.slice(3).trim();
  if (text.startsWith("# ")) return text.slice(2).trim();
  return text;
}

export function parseTemplateStructure(
  text: string,
  options?: { deliverableId?: string; preferCreac?: boolean; excerptChars?: number },
): TemplateSection[] {
  const raw = (text || "").trim();
  if (!raw) return [];

  const deliverableId = options?.deliverableId ?? "";
  const useCreac =
    options?.preferCreac ??
    ["aos-discretionary-brief", "aos_discretionary_brief"].includes(deliverableId);
  const excerptChars = options?.excerptChars ?? 600;

  const lines = raw.replace(/\r\n/g, "\n").split("\n");
  const headingIdxs: number[] = [];
  for (let i = 0; i < lines.length; i++) {
    if (isHeadingLine(lines[i] ?? "")) headingIdxs.push(i);
  }

  if (!headingIdxs.length) {
    return [
      {
        id: slug("Full template", 0),
        label: "Full template",
        role: "other",
        contentExcerpt: raw.slice(0, excerptChars),
        order: 0,
      },
    ];
  }

  const sections: TemplateSection[] = headingIdxs.map((start, order) => {
    const end = headingIdxs[order + 1] ?? lines.length;
    const label = headingLabel(lines[start] ?? "");
    const body = lines.slice(start + 1, end).join("\n").trim();
    return {
      id: slug(label, order),
      label,
      role: useCreac ? detectCreacRole(label) : "other",
      contentExcerpt: (body || label).slice(0, excerptChars),
      order,
    };
  });

  if (useCreac) {
    const conclusionIdxs = sections
      .map((s, i) => (s.role === "conclusion" ? i : -1))
      .filter((i) => i >= 0);
    if (conclusionIdxs.length >= 2) {
      const last = conclusionIdxs[conclusionIdxs.length - 1]!;
      sections[last]!.role = "conclusion_close";
    }
  }

  return sections;
}

/** Default CREAC outline when no firm upload yet (AOS discretionary brief). */
export const DEFAULT_AOS_CREAC_SECTIONS: TemplateSection[] = [
  {
    id: "conclusion-open-0",
    label: "Conclusion (opening)",
    role: "conclusion",
    contentExcerpt:
      "State the requested relief up front (e.g. grant AOS / favorable exercise of discretion).",
    order: 0,
  },
  {
    id: "rule-1",
    label: "Rule",
    role: "rule",
    contentExcerpt:
      "Preserve the firm template’s statement of law (INA §245(a), PM-602-0199, Matter of Marin / Patel).",
    order: 1,
  },
  {
    id: "explanation-2",
    label: "Explanation",
    role: "explanation",
    contentExcerpt: "Explain how the rule operates — totality of the circumstances; AOS as administrative grace.",
    order: 2,
  },
  {
    id: "analysis-3",
    label: "Analysis",
    role: "analysis",
    contentExcerpt:
      "Apply matter facts (hardship, equities, qualifying relative, adverse factors) against the preserved Rule.",
    order: 3,
  },
  {
    id: "conclusion-close-4",
    label: "Conclusion (closing)",
    role: "conclusion_close",
    contentExcerpt: "Restate the request for relief guided by the Analysis outcome.",
    order: 4,
  },
];

export const CREAC_ROLE_LABELS: Record<CreacRole, string> = {
  conclusion: "Conclusion (open)",
  rule: "Rule — preserve from template",
  explanation: "Explanation — preserve / light tweak",
  analysis: "Analysis — matter facts",
  conclusion_close: "Conclusion (close)",
  introduction: "Introduction",
  caption: "Caption",
  other: "Other",
};

/** AOS drafting fact field → CREAC slot. */
export const AOS_FACT_CREAC_MAP: Record<string, CreacRole> = {
  applicantName: "caption",
  aNumber: "caption",
  clientStatus: "analysis",
  entryDate: "analysis",
  portOfEntry: "analysis",
  entryVisaType: "analysis",
  petitionerName: "analysis",
  petitionerRelationship: "analysis",
  qualifyingRelative: "analysis",
  i130ApprovedDate: "analysis",
  i485FiledDate: "analysis",
  caseTheme: "conclusion",
  caseThemeBrief: "conclusion_close",
  sectionAHeading: "analysis",
  sectionAFacts: "analysis",
  sectionBHeading: "analysis",
  sectionBFacts: "analysis",
  adverseHeading: "analysis",
  adverseFacts: "analysis",
  adverseFactorBrief: "analysis",
  adverseFactors: "analysis",
  positiveEquities: "analysis",
  balancingInventory: "analysis",
  departureHarm: "explanation",
  extremeHardshipFactors: "analysis",
  inadmissibilityGrounds: "analysis",
  priorFilings: "analysis",
  supportingDocs: "analysis",
  reliefSought: "conclusion",
};

export const CLASSIFICATION_LABELS: Record<string, string> = {
  PRESERVE: "PRESERVE",
  FILL: "FILL",
  CAPTION: "CAPTION",
  BOILERPLATE: "BOILERPLATE",
};

export function classificationBadgeClass(classification: string): string {
  switch (classification) {
    case "PRESERVE":
      return "bg-indigo-50 text-indigo-900 ring-indigo-600/25";
    case "FILL":
      return "bg-amber-50 text-amber-950 ring-amber-600/25";
    case "CAPTION":
      return "bg-sky-50 text-sky-900 ring-sky-600/25";
    case "BOILERPLATE":
      return "bg-slate-100 text-slate-800 ring-slate-500/25";
    default:
      return "bg-slate-50 text-slate-700 ring-slate-500/20";
  }
}

export function roleBadgeClass(role: CreacRole): string {
  switch (role) {
    case "rule":
      return "bg-indigo-50 text-indigo-800 ring-indigo-600/20";
    case "explanation":
      return "bg-violet-50 text-violet-800 ring-violet-600/20";
    case "analysis":
      return "bg-amber-50 text-amber-900 ring-amber-600/20";
    case "conclusion":
    case "conclusion_close":
      return "bg-emerald-50 text-emerald-800 ring-emerald-600/20";
    case "introduction":
      return "bg-sky-50 text-sky-800 ring-sky-600/20";
    default:
      return "bg-slate-50 text-slate-700 ring-slate-500/20";
  }
}

File: services/api/app/services/template_structure.py
"""Parse firm template text into structural outline (headings / CREAC)."""

from __future__ import annotations

import re
from typing import Any, Literal

CreacRole = Literal[
    "conclusion",
    "rule",
    "explanation",
    "analysis",
    "conclusion_close",
    "introduction",
    "caption",
    "other",
]

_CREAC_PATTERNS: list[tuple[re.Pattern[str], CreacRole]] = [
    (re.compile(r"^\s*(i+\.?\s+)?conclusion\b", re.I), "conclusion"),
    (re.compile(r"^\s*(v+\.?\s+)?conclusion\b", re.I), "conclusion"),
    (re.compile(r"^\s*(ii+\.?\s+)?(legal\s+)?standard\b", re.I), "rule"),
    (re.compile(r"^\s*(ii+\.?\s+)?rule\b", re.I), "rule"),
    (re.compile(r"^\s*(applicable\s+)?law\b", re.I), "rule"),
    (re.compile(r"^\s*(iii+\.?\s+)?explanation\b", re.I), "explanation"),
    (re.compile(r"^\s*(discussion\s+of\s+(the\s+)?law|legal\s+framework)\b", re.I), "explanation"),
    (re.compile(r"^\s*(iv+\.?\s+)?(argument|analysis|application|discussion)\b", re.I), "analysis"),
    (re.compile(r"^\s*(totality|discretionary\s+factors|positive\s+equities)\b", re.I), "analysis"),
    (re.compile(r"^\s*(i+\.?\s+)?introduction\b", re.I), "introduction"),
    (re.compile(r"^\s*(caption|in\s+the\s+matter\s+of)\b", re.I), "caption"),
    (re.compile(r"^\s*(certificate\s+of\s+service|table\s+of\s+contents)\b", re.I), "other"),
]

_NUMBERED_HEADING = re.compile(
    r"^\s*((?:[IVXLC]+\.|[A-Z]\.|§?\d+(?:\.\d+)*\.?|[0-9]+(?:\.[0-9]+)*\.?)\s+)(.{3,120})$"
)
_ALL_CAPS_HEADING = re.compile(r"^[A-Z0-9][A-Z0-9\s\-–—,.'()]{2,100}$")
_MARKDOWN_HEADING = re.compile(r"^#{1,3}\s+(.+)$")


def _slug(label: str, order: int) -> str:
    base = re.sub(r"[^a-z0-9]+", "-", label.lower()).strip("-")[:48] or "section"
    return f"{base}-{order}"


def detect_creac_role(label: str) -> CreacRole:
    cleaned = (label or "").strip()
    # Strip leading roman / number markers for matching
    stripped = re.sub(r"^(?:[IVXLC]+\.|[A-Z]\.|§?\d+(?:\.\d+)*\.?)\s*", "", cleaned, flags=re.I)
    for pattern, role in _CREAC_PATTERNS:
        if pattern.search(stripped) or pattern.search(cleaned):
            return role
    return "other"


def _is_heading_line(line: str) -> bool:
    text = line.strip()
    if not text or len(text) > 140:
        return False
    if _MARKDOWN_HEADING.match(text):
        return True
    if text.startswith("## ") or text.startswith("# "):
        return True
    if _NUMBERED_HEADING.match(text):
        return True
    # ALL CAPS short line (exclude sentences with many lowercase leftovers already filtered)
    if _ALL_CAPS_HEADING.match(text) and " " in text and not text.endswith("."):
        letters = [c for c in text if c.isalpha()]
        if letters and all(c.isupper() for c in letters) and len(letters) >= 4:
            return True
    # CREAC-ish bare labels
    if detect_creac_role(text) != "other" and len(text.split()) <= 8:
        return True
    return False


def _heading_label(line: str) -> str:
    text = line.strip()
    md = _MARKDOWN_HEADING.match(text)
    if md:
        return md.group(1).strip()
    if text.startswith("## "):
        return text[3:].strip()
    if text.startswith("# "):
        return text[2:].strip()
    return text


def parse_template_structure(
    text: str,
    *,
    deliverable_id: str = "",
    prefer_creac: bool | None = None,
    excerpt_chars: int = 600,
) -> list[dict[str, Any]]:
    """
    Split template text into ordered sections.

    Returns list of:
      {id, label, role, contentExcerpt, order}
    """

    raw = (text or "").strip()
    if not raw:
        return []

    use_creac = prefer_creac
    if use_creac is None:
        use_creac = deliverable_id in {"aos-discretionary-brief", "aos_discretionary_brief"}

    lines = raw.replace("\r\n", "\n").split("\n")
    # Find heading indices
    heading_idxs: list[int] = []
    for i, line in enumerate(lines):
        if _is_heading_line(line):
            heading_idxs.append(i)

    sections: list[dict[str, Any]] = []
    if not heading_idxs:
        # Single blob — still try to surface CREAC labels inside paragraphs
        role: CreacRole = "other"
        label = "Full template"
        if use_creac and re.search(r"\brule\b|\blegal\s+standard\b", raw, re.I):
            label = "Unstructured template (scan for CREAC labels)"
        sections.append(
            {
                "id": _slug(label, 0),
                "label": label,
                "role": role,
                "contentExcerpt": raw[:excerpt_chars],
                "order": 0,
            }
        )
        return sections

    for order, start in enumerate(heading_idxs):
        end = heading_idxs[order + 1] if order + 1 < len(heading_idxs) else len(lines)
        label = _heading_label(lines[start])
        body = "\n".join(lines[start + 1 : end]).strip()
        role = detect_creac_role(label) if use_creac else "other"
        if not use_creac:
            role = "other"
        sections.append(
            {
                "id": _slug(label, order),
                "label": label,
                "role": role,
                "contentExcerpt": (body or label)[:excerpt_chars],
                "order": order,
            }
        )

    # If prefer CREAC and we got multiple conclusions, mark last as conclusion_close
    if use_creac:
        conclusion_idxs = [i for i, s in enumerate(sections) if s["role"] == "conclusion"]
        if len(conclusion_idxs) >= 2:
            last = conclusion_idxs[-1]
            sections[last]["role"] = "conclusion_close"
        # If we detected almost no CREAC roles, keep heading tree as-is (roles mostly other)

    return sections


def format_structure_for_prompt(
    sections: list[dict[str, Any]],
    *,
    max_chars: int = 3500,
) -> str:
    """Render TEMPLATE STRUCTURE block for drafting prompts."""

    if not sections:
        return ""
    lines = [
        "## TEMPLATE STRUCTURE (CREAC / outline)",
        "Preserve RULE and EXPLANATION wording from the firm template. "
        "Map new matter facts into ANALYSIS. Conclusions should follow from that analysis.",
        "",
    ]
    used = 0
    for sec in sections:
        role = str(sec.get("role") or "other")
        label = str(sec.get("label") or "Section")
        excerpt = str(sec.get("contentExcerpt") or "").strip()
        preserve = role in {"rule", "explanation"}
        tag = "PRESERVE from template" if preserve else (
            "FILL with matter facts" if role == "analysis" else
            "Guide from analysis outcome" if role in {"conclusion", "conclusion_close"} else
            "Adapt as needed"
        )
        block = f"### [{role.upper()}] {label}\n- Instruction: {tag}\n"
        if excerpt:
            block += f"- Template excerpt:\n{excerpt}\n"
        block += "\n"
        if used + len(block) > max_chars:
            lines.append("(additional sections truncated)")
            break
        lines.append(block)
        used += len(block)
    return "\n".join(lines).strip()


def creac_fact_slot_for_field(field_id: str) -> CreacRole | None:
    """Map structured drafting fact field ids to CREAC slots (AOS)."""

    analysis_fields = {
        "qualifyingRelative",
        "extremeHardshipFactors",
        "adverseFactors",
        "adverseFacts",
        "adverseHeading",
        "adverseFactorBrief",
        "positiveEquities",
        "balancingInventory",
        "inadmissibilityGrounds",
        "priorFilings",
        "clientStatus",
        "supportingDocs",
        "entryDate",
        "portOfEntry",
        "entryVisaType",
        "petitionerName",
        "petitionerRelationship",
        "i130ApprovedDate",
        "i485FiledDate",
        "sectionAHeading",
        "sectionAFacts",
        "sectionBHeading",
        "sectionBFacts",
    }
    conclusion_fields = {"reliefSought", "caseTheme", "caseThemeBrief"}
    caption_fields = {"applicantName", "aNumber"}
    explanation_fields = {"departureHarm"}
    if field_id in analysis_fields:
        return "analysis"
    if field_id in conclusion_fields:
        return "conclusion"
    if field_id in caption_fields:
        return "caption"
    if field_id in explanation_fields:
        return "explanation"
    return None


def format_creac_facts_for_prompt(drafting_fields: dict[str, Any] | None) -> str:
    """Group matter facts under CREAC slots for ANALYSIS vs conclusion guidance."""

    if not drafting_fields:
        return ""
    analysis: list[str] = []
    conclusions: list[str] = []
    other: list[str] = []
    for key, val in drafting_fields.items():
        if not val:
            continue
        if isinstance(val, list):
            rendered = "; ".join(str(v).strip() for v in val if str(v).strip())
        else:
            rendered = str(val).strip()
        if not rendered:
            continue
        slot = creac_fact_slot_for_field(key)
        line = f"- {key}: {rendered}"
        if slot == "analysis":
            analysis.append(line)
        elif slot == "conclusion":
            conclusions.append(line)
        else:
            other.append(line)
    if not (analysis or conclusions or other):
        return ""
    parts = [
        "## FACTS FOR ANALYSIS (map into CREAC Analysis; do not overwrite Rule)",
        "Use these matter facts to show how the case aligns with the Rule preserved above.",
    ]
    if analysis:
        parts.append("### Analysis slot")
        parts.extend(analysis)
    if conclusions:
        parts.append("### Conclusion guidance")
        parts.extend(conclusions)
    if other:
        parts.append("### Other facts")
        parts.extend(other)
    return "\n".join(parts)
